### ISSUE TITLE
[Bug/Enhancement] ErrorOr should not "invent" a result value in case error, it should force caller to handle error instead

### ISSUE THREAD

**Issue by duongphuhiep** [amantinband/error-or#92]
ErrorOr is following the [Golang-style] (and it is Bad) 👇
```C#
ErrorOr<float> result = Divide(4, 0);

// I forgot to handle error here and it is the "root cause" for future troubles
// if (result.IsError) { ... }

// thing continue to work with the hidden error(s) in the result.
var nextResult = result.Value * 2;  //nextResult = 0

//... until the application crashed somewhere far away from the "root cause"
// and it is hard to trace back to the "root cause".
```

My suggestion: ErrorOr should follow the [Rust-style] instead 👇
```C#
ErrorOr<float> result = Divide(4, 0);

// I forgot to handle error here
// if (result.IsError) { ... }

// I try to get the result (without error handling)
var nextResult = result.Unwrap() * 2;  // the Unwrap function will "crash" / "panic" here
```

The first example shows that ErrorOr is "inventing" a false result (result.Value = 0) and it is not logical. You can also see it as a buggy behavior. 

Logically, accessing to the `result.Value` should crashed because the value for `Divide(4, 0)` simply doesn't exist.

It is what the [Rust-style] is doing and I demonstrated it in the second example with the imaginary `Unwrap` function.

In other words, the `Result<T,E>` of Rust beautifuly "forced" developers to handle error: In order to Get the result T you MUST to explicitly handle the "E" one way or other: in a `match` statement or by a simple `unwrap()` function to get the T which will panic the E... 

[Golang-style]: https://go.dev/blog/error-handling-and-go
[Rust-style]: https://doc.rust-lang.org/rust-by-example/error/result.html

**Comment by lucafabbri** [comment]
Hi Duong,

I'be been using ErrorOr for a while and I guess you are missing the whole picture.
ErrorOr is an actual implementation with ROP (Railway Oriented Programming) in mind.

Direct access to Value is not ideal as well as throwing an Exception would invalidate the effort of managing two "rails" (success, error) while managing a flow.

In a common flow like yours you should not miss to handle errors.

Otherwise you can simply use a more ROP syntax like this:

`return Divide(4,0).Then(value => value * 2).Then(nextResult => ... );`

This avoids any issue with unconsistents values, it is more concise and readable.

Cheers,

**Comment by lucafabbri** [comment]
@duongphuhiep 

Hereafter another solution

```
public static class ErrorOrExtension
{
    public static TValue? ValueOrNull<TValue>(this ErrorOr<TValue> result)
    {
        return result.IsError ? default : result.Value;
    }
}
```

**Comment by AkosLukacs** [comment]
I feel this is a mix of three things (I just found this while searching for a good Result library, so I'm not a long time user, but used F# for quite some time): 
 - In the case of a value type, ErrorOr doesn't "invent" the value, a value type like float simply always have a default value. If it was a class / record, it would have been `null`.
 - IMO the documentation should show the "right" way to use it. Now the readme first shows the `if (result.IsError) { ...` check, and directly accesses the `result.Value` as well. Not ideal. Show the "pit of success" path for new users!
 - And the design issue of allowing easy access to `.Value` without check. In most cases with reference types the next thing after accessing `.Value` in the error case will be a NullReferenceException. One way would be to change the behaviour of `.Value` to throw right away, if we are in the Error case? Since it's a property anyways. And call it `.ValueOrThrow`? Or just remove that from the public API and force a `Match(...)` call on the `Result`? Maybe this intention comes from an ease of use in the first time perspective, but can be misleading for newcomers who see the example.


@lucafabbri the example in your post yesterday would behave exactly same for a float, since the `default` of a float is `0`:

```csharp
Console.WriteLine($"default(float): {default(float)}");
var result  = Dummy.Divide(4, 0);
Console.WriteLine($"result.Value: {result .Value}");
Console.WriteLine($"result.Value is default(float): {result.Value is default(float)}");
```
Output:
```
default(float): 0
result.Value: 0
result.Value is default(float): True
```

And it is `default` by default in the code: https://github.com/amantinband/error-or/blob/68c8a06ea953b4e1f69aed170504b69ff637535c/src/ErrorOr.cs#L8

**Comment by duongphuhiep** [comment]
hi, thanks @lucafabbri I'm aware of Railway Oriented Programming, and different way to do things in ErrorOr. I'm only pointing out the default in the current api like @AkosLukacs said.

My suggestions are:

- rename `.Value` to `.ValueOrDefault` to explicitely tell caller that this property would return a default value in case the value doesn't exist. (I don't like this `.ValueOrDefault` but it is still better)

- replace `.Value` with `.ValueOrThrow` which would throw exception when you try to access the value from the rail error branch.

- or remove `.Value` completely if we think that direct access to Value is bad (I have a different opinion on this but we will go with your recommendation)

**Comment by lucafabbri** [comment]
@AkosLukacs @duongphuhiep being honest with you the access to Value has been discussed many times with my team as some them were accessing it out of ROP and without checking is an error and this was leading to inconsistent output.
For this reason we implemented a .ValueOrNull() extension.

Thing is that when you are going to reed the Value out ROP, you still not want to throw an exception (avoid exception handling is what the library is made for) so I'd rather go for a solution that falls back to null.

This will be a breaking change in the current implementation of the library, so I'd rather ask the @amantinband opinion

**Comment by amantinband** [comment]
Enforcement #105. Will be available as part of ErrorOr 3.x
